en/services/mission.md

```diff
@@ -149,17 +144,33 @@ If sent in messages `float` parameter fields the value should be sent as-is.
 
 ## Param 5, 6 For Non-Positional Data
 
-Param5, param6, param7 may also be used for non-positional information.
+`Param5`, `param6`, `param7` may also be used for non-positional information.
 In this case the [MISSION_ITEM_INT.frame](#MISSION_ITEM_INT) should be set to [MAV_FRAME_MISSION](../messages/common.md#MAV_FRAME_MISSION) (this is equivalent to say "the frame data is irrelevant").
 
 As param5 and param6 are sent in *integer* fields, generally you should design mission items/MAV_CMDs such that these only include integer data (and are sent as-is/unscaled).
 If these must be used for real numbers and scaling is required, then this must be noted in the mission item itself.
 
-
 ## Operations {#operations}
 
 This section defines all the protocol operations.
 
+### Detecting Mission/Plan Changes
+
+Mission upload and download can be expensive operations, in particular for large missions.
+A GCS can avoid unnecessary uploads and downloads by first checking whether it has a matching id(s) for the corresponding plan on the vehicle.
+
+The current id for different parts of the plan are streamed in [MISSION_CURRENT](#MISSION_CURRENT) message, using the appropriately named fields: `mission_id`, `fence_id`, `rally_points_id`.
+These values are `0` if there is no plan uploaded, or if detecting plan changes is not supported by the flight stack.
+
+The IDs are generated by the flight stack when a new mission, rally point, or geofence, is _uploaded_ to the vehicle (at which point it also starts publishing the value in [MISSION_CURRENT](#MISSION_CURRENT)).
+MAVLink puts no particular requirements on _how_ the "opaque_id" values are calculated by the flight stack (this is why they are named "opaque").
+The only expectation is that the scheme used makes it unlikely that the GCS will incorrectly determine that it has a matching plan (a flight stack might use file hashes, plan checksums, pseudo-random numbers, or some other technique).
+
+On upload, the generated ID is sent to the GCS in the final part of the upload sequence in the [MISSION_ACK.opaque_id](#MISSION_ACK) field.
+On download, the stored ID is sent to the GCS in the [MISSION_COUNT.opaque_id](#MISSION_COUNT) field.
+
+The GCS should store the value of the ID from the flight stack as the "current id" for whatever part of the plan was uploaded/download.
+It can then monitor `MISSION_CURRENT`, and check its cached values against the current plan ids to determine whether it has a matching mission, or needs to download the mission from the vehicle.
```

Comment:
@julianoes @peterbarker This is the mission id stuff. Can you sanity check this bit matches expectation?